### PR TITLE
Feature/create giveaway

### PR DESCRIPTION
**Pull request by quocthinhle**
Feature giveaway:

> !giveaway:{quantity} >> hh:mm DD/MM/YYYY >> {message}

### CODE REVIEW

File: src/core/common/constant/command.constant.js [hunk 1]
```diff
@@ -1,2 +1,4 @@
 export const COMMAND_PREFIX = '!';
 export const TITLE_VALUE_SEPARATOR = '>>';
+export const DATE_FORMAT = 'hh:mm DD/MM/YYYY';
+export const DEFAULT_TIME = '23:59';
```
Comment (ichhoa129) [inline src/core/common/constant/command.constant.js]: DEFAULT_TIME for what?

File: src/core/common/constant/command.constant.js [hunk 2]
```diff
@@ -1,2 +1,4 @@
 export const COMMAND_PREFIX = '!';
 export const TITLE_VALUE_SEPARATOR = '>>';
+export const DATE_FORMAT = 'hh:mm DD/MM/YYYY';
```
Comment (imrim12) [inline src/core/common/constant/command.constant.js]: Should be in another file like date.constant.js

File: src/core/config/discord.config.js [hunk 3]
```diff
@@ -8,7 +8,15 @@ class DiscordConfig {
...
     constructor() {
-        this.client = new Client({ intents: [BOT_INTENTS.GUILDS, BOT_INTENTS.GUILD_MESSAGES, BOT_INTENTS.DIRECT_MESSAGES] });
+        this.client = new Client({ 
+            intents: [
+                BOT_INTENTS.GUILDS,
+                BOT_INTENTS.GUILD_MESSAGES,
+                BOT_INTENTS.DIRECT_MESSAGES,
+                BOT_INTENTS.GUILD_MESSAGE_REACTIONS,
+            ],
+            partials: ['MESSAGE', 'CHANNEL', 'REACTION'],
```
Comment (imrim12) [inline src/core/config/discord.config.js]: Inconsistency in code, use BOT_INTENTS enum above but use raw string here

File: src/core/modules/give-away/give-away.request-process.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import { TITLE_VALUE_SEPARATOR, DATE_FORMAT, DEFAULT_TIME } from 'core/common/constant';
+import moment from 'moment';
+
+export class GiveAwayRequestProcess {
+    constructor(content) {
+        this.content = content;
+    }
+
+    seperateData() {
+        let arr = this.content.split(TITLE_VALUE_SEPARATOR).map(e => e.trim());
```
Comment (imrim12) [inline src/core/modules/give-away/give-away.request-process.js]: inconsistency in code, use const for non-reassigned value

**Review by imrim12 — DISMISSED** [review]

**Review by imrim12 — DISMISSED** [review]

File: src/core/modules/give-away/give-away.request-process.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        if (dateDetail.length == 1) {
+            dateDetail = [DEFAULT_TIME, ...dateDetail];
+        }
+        return dateDetail;
+    }
+
+    validateQuantity(quantity) {
+        let number = parseInt(quantity);
+        if (Number.isNaN(number)) return false;
+        return number > 0;
```
Comment (ichhoa129) [inline src/core/modules/give-away/give-away.request-process.js]: You can use Joi to validate the payload
https://joi.dev/api/

File: src/core/modules/give-away/give-away.request-process.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import moment from 'moment';
+
+export class GiveAwayRequestProcess {
+
+    constructor(content) {
+        this.content = content;
+    }
+
+    seperateData() {
+        let arr = this.content.split(TITLE_VALUE_SEPARATOR).map(e => e.trim());
```
Comment (ichhoa129) [inline src/core/modules/give-away/give-away.request-process.js]: Array of what ? you should name it clearly

File: src/core/modules/give-away/give-away.request-process.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    validateQuantity(quantity) {
+        let number = parseInt(quantity);
+        if (Number.isNaN(number)) return false;
+        return number > 0;
+    }
+
+    validateDate(date) {
+        const dateWithFormat = moment(date, DATE_FORMAT);
```
Comment (ichhoa129) [inline src/core/modules/give-away/give-away.request-process.js]: formatedDate should be better

File: src/core/modules/give-away/give-away.request-process.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        const now = moment(new Date());
+        if (!dateWithFormat.isValid()) return null;
+        if (dateWithFormat - now < 0) return null;
```
Comment (ichhoa129) [inline src/core/modules/give-away/give-away.request-process.js]: Use one if statement for both condition

File: src/core/modules/give-away/give-away.schedule.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            return;
+        }
+        let currentIndex = 0;
+        let timeStamp = 0;
+        const timer = setInterval(() => {
+            if (timeStamp % 2 == 1) {
+                this.channel.send("⌛");
+            } else {
+                this.channel.send("⏳");
+            }
```
Comment (ichhoa129) [inline src/core/modules/give-away/give-away.schedule.js]: Hmm, what is this?

File: src/core/modules/give-away/give-away.schedule.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+    async replyMessage(luckyUsers) {
+        this.message.reply(`Giveaway start`);
+        this.channel.send(`Guess who will be the luckiests?`);
+        if (luckyUsers.length == 0) {
+            this.message.reply(`There is actually 0 person reacted to your message!`);
+            return;
+        }
+        let currentIndex = 0;
+        let timeStamp = 0;
```
Comment (ichhoa129) [inline src/core/modules/give-away/give-away.schedule.js]: You should name it timeCounting or something easier to understand

**Review by ichhoa129 — CHANGES_REQUESTED** [review]

File: src/core/common/constant/command.constant.js (hunk 1 above)
Comment (quocthinhle) [reply src/core/common/constant/command.constant.js]: It's like when an event is scheduled without specifying time, just only date.
(!giveaway:1 >> 31/12/2021 >> "ABC" => 31/12/2021 -> 23:59:00 31/12/2021)

File: src/core/modules/give-away/give-away.request-process.js (hunk 8 above)
Comment (quocthinhle) [reply src/core/modules/give-away/give-away.request-process.js]: I'll fix it, thx u

File: src/core/modules/give-away/give-away.schedule.js (hunk 9 above)
Comment (quocthinhle) [reply src/core/modules/give-away/give-away.schedule.js]: It works like this 
image

File: src/core/modules/give-away/give-away.request-process.js (hunk 8 above)
Comment (quocthinhle) [reply src/core/modules/give-away/give-away.request-process.js]: Should I delete all the previous commits?
Just keep the current one?

File: src/core/modules/give-away/give-away.schedule.js (hunk 9 above)
Comment (ichhoa129) [reply src/core/modules/give-away/give-away.schedule.js]: Hmm, the 2 icons are different, LOL

File: src/core/modules/give-away/give-away.request-process.js (hunk 8 above)
Comment (ichhoa129) [reply src/core/modules/give-away/give-away.request-process.js]: Just push more commit, or squash into one commit and force push

File: src/core/common/constant/command.constant.js (hunk 1 above)
Comment (ichhoa129) [reply src/core/common/constant/command.constant.js]: So you should name it clearly, like DEFAUT_SCHEDULED_TIME and put it in other file, giveaway.constant.js

**Review by ichhoa129 — APPROVED** [review]

**Pull request merged**